From d50ab1579abfa28bf4f2c5eefd78fcd45b320c0d Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@st.com>
Date: Wed, 30 Nov 2022 15:08:20 +0100
Subject: [PATCH 303/529] ARM: dts: stm32: add cpufreq support on stm32mp13

This patch adds support for cpufreq with
thermal monitoring on stm32mp13.
The CPU opp are set by OPTEE.
Above 95 degrees the OPP is limited to the lowest one.
Above 120 degrees a thermal shutdown is initiated to
protect the SOC.

Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/279691
Change-Id: I51517eef0861e98a2e5850555ec5a9adf8e6e3fb
Signed-off-by: Pascal Paillet <p.paillet@st.com>
---
 arch/arm/boot/dts/stm32mp131.dtsi    | 49 ++++++++++++++++++++++++++++
 arch/arm/boot/dts/stm32mp135f-dk.dts |  4 +++
 2 files changed, 53 insertions(+)

--- a/arch/arm/boot/dts/stm32mp131.dtsi
+++ b/arch/arm/boot/dts/stm32mp131.dtsi
@@ -7,6 +7,7 @@
 #include <dt-bindings/clock/stm32mp13-clks.h>
 #include <dt-bindings/regulator/st,stm32mp13-regulator.h>
 #include <dt-bindings/reset/stm32mp13-resets.h>
+#include <dt-bindings/thermal/thermal.h>
 
 / {
 	#address-cells = <1>;
@@ -20,6 +21,11 @@
 			compatible = "arm,cortex-a7";
 			device_type = "cpu";
 			reg = <0>;
+			clocks = <&scmi_perf 0>;
+			clock-names = "cpu";
+			nvmem-cells = <&part_number_otp>;
+			nvmem-cell-names = "part_number";
+			#cooling-cells = <2>;
 		};
 	};
 
@@ -43,6 +49,11 @@
 			linaro,optee-channel-id = <0>;
 			shmem = <&scmi_shm>;
 
+			scmi_perf: protocol@13 {
+				reg = <0x13>;
+				#clock-cells = <1>;
+			};
+
 			scmi_clk: protocol@14 {
 				reg = <0x14>;
 				#clock-cells = <1>;
@@ -120,6 +131,34 @@
 		};
 	};
 
+	thermal-zones {
+		cpu_thermal: cpu-thermal {
+			polling-delay-passive = <0>;
+			polling-delay = <0>;
+			thermal-sensors = <&dts>;
+
+			trips {
+				cpu_alert: cpu-alert0 {
+					temperature = <95000>;
+					hysteresis = <10000>;
+					type = "passive";
+				};
+				cpu_crit: cpu-crit0 {
+					temperature = <120000>;
+					hysteresis = <0>;
+					type = "critical";
+				};
+			};
+
+			cooling-maps {
+				map0 {
+					trip = <&cpu_alert>;
+					cooling-device = <&cpu0 1 1>;
+				};
+			};
+		};
+	};
+
 	soc {
 		compatible = "simple-bus";
 		#address-cells = <1>;
@@ -1265,6 +1304,16 @@
 			};
 		};
 
+		dts: thermal@50028000 {
+			compatible = "st,stm32-thermal";
+			interrupts = <GIC_SPI 124 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&rcc DTS>;
+			clock-names = "pclk";
+			#thermal-sensor-cells = <0>;
+			reg = <0x50028000 0x100>;
+			status = "disabled";
+		};
+
 		hdp: hdp@5002a000 {
 			compatible = "st,stm32mp1-hdp";
 			reg = <0x5002a000 0x400>;
--- a/arch/arm/boot/dts/stm32mp135f-dk.dts
+++ b/arch/arm/boot/dts/stm32mp135f-dk.dts
@@ -128,6 +128,10 @@
 	};
 };
 
+&dts {
+	status = "okay";
+};
+
 &i2c1 {
 	pinctrl-names = "default", "sleep";
 	pinctrl-0 = <&i2c1_pins_a>;
